pulsar-io/jdbc/src/main/java/org/apache/pulsar/io/jdbc/JdbcAbstractSink.java

```diff
@@ -0,0 +1,190 @@
+        String username = jdbcSinkConfig.getUserName();
+        String password = jdbcSinkConfig.getPassword();
+        if (username != null) {
+            properties.setProperty("user", username);
+        }
+        if (password != null) {
+            properties.setProperty("password", password);
+        }
+
+        connection = JdbcUtils.getConnection(jdbcUrl, properties);
+        log.info("Connection opened");
+
+        schema = jdbcSinkConfig.getSchema();
+        tableName = jdbcSinkConfig.getTableName();
+        tableId = JdbcUtils.getTableId(connection, tableName);
+        tableDefinition = JdbcUtils.getTableDefinition(connection, tableId);
+        insertStatement = JdbcUtils.buildInsertStatement(connection, JdbcUtils.buildInsertSql(tableDefinition));
+
+        timeoutMs = jdbcSinkConfig.getTimeoutMs();
+        batchSize = jdbcSinkConfig.getBatchSize();
+        incomingList = Lists.newArrayList();
+        swapList = Lists.newArrayList();
+        isFlushing = new AtomicBoolean(false);
+
+        flushExecutor = Executors.newScheduledThreadPool(1);
+        flushExecutor.scheduleAtFixedRate(() -> flush(), timeoutMs, timeoutMs, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (!connection.getAutoCommit()) {
+            connection.commit();
+        }
+        flushExecutor.shutdown();
+        if (connection != null) {
+            connection.close();
+        }
+        log.info("Connection Closed");
+    }
+
+    @Override
+    public void write(Record<T> record) throws Exception {
+        int number;
+        synchronized (incomingList) {
+            incomingList.add(record);
+            number = incomingList.size();
+        }
+
+        if (number == batchSize) {
+            flushExecutor.schedule(() -> flush(), 0, TimeUnit.MILLISECONDS);
+        }
+    }
+
+    // bind value with a PreparedStetement
+    public abstract void bindValue(
+        PreparedStatement statement,
+        Record<T> message) throws Exception;
+
+
+    private void flush() {
+        // if not in flushing state, do flush, else return;
+        if (isFlushing.compareAndSet(false, true)) {
+            log.debug("Starting flush, queue size: {}", incomingList.size());
+            checkState(swapList.isEmpty(),
+                "swapList should be empty since last flush. swapList.size: " + swapList.size());
+
+            synchronized (incomingList) {
+                List<Record<T>> tmpList;
+                swapList.clear();
+
+                tmpList = swapList;
+                swapList = incomingList;
+                incomingList = tmpList;
+            }
+
+            int updateCount = 0;
+            boolean noInfo = false;
+            try {
+                // bind each record value
+                for (Record<T> record : swapList) {
+                    bindValue(insertStatement, record);
+                    insertStatement.addBatch();
+                    record.ack();
```

Comment:
don't ack here. ack when connection commit. this way it is  guaranteed the records are written to database

Reply:
@jiazhai it seems that this is not changed yet?